### ISSUE TITLE
v0.3.1 GenericStorage: cost of energy lost by self-discharge are not taken into account in objective value

### ISSUE THREAD

**Issue by baumconsult** [oemof/oemof#615]
[storage_check_2.txt](https://github.com/oemof/oemof/files/3301732/storage_check_2.txt)
[storage_analysis.xlsx](https://github.com/oemof/oemof/files/3301734/storage_analysis.xlsx)

I used storage_check_2.py (see attached text file) for checking the GenericStorage module in v0.3.1. By varying the parameters, I found that the cost of energy lost by self-discharge are not taken into account in the objective value, while all other costs are taken into account. The parameter settings in the attached example lead to total system costs of 81.79 units, while the objective value 73.31 units. If the self-discharge is set to zero, the total system costs always equal the objective value.

In conclusion, the GenericStorage module of v0.3.1 does not calculate the optimum system configuration if a storage with non-zero self-discharge is included. In such a case, to much use of the storage (dispatch calculation) or a too big storage (investment calculation) is obtained as apparently optimum solution.

**Comment by uvchik** [comment]
I do not understand the general idea off adding loss costs. Electricity losses produce costs because one has to produce more energy to satisfy the demand but I would assign extra costs for the losses.

**Comment by baumconsult** [comment]
Attributing costs of lost energy explicitely to a component allows easier modelling of situations in which different flexibility options are compared which deal equally well with a rigid generation and a rigid demand, but which have different costs. The first situation of this kind which we studied was https://github.com/baumconsult/gridcon where we identified the optimum combination of a grid connection and local energy storage. Both options have investment costs and costs of lost electricity depending on the way how they are used.
Anyway, it is possible to attribute variable costs to incoming and outcoming flows in the GenericStorage module. This is simply incomplete as no costs can be attributed to the "internal flow" of self-discharged energy.
It would be great if you could modify the GenericStorage module in such a way that variable costs can also be attributed to self-discharged energy.

**Comment by jakob-wo** [comment]
Hi everyone,

I like the idea of questioning and double checking the pricing of the losses.
After having a look at the example I would say oemof DOES consider the costs of temporal losses (self-discharing) and the calculated objective value is correct.
This opinion comes from the following observations:

- The `GenericStorage` is `balanced=True`, therefore the SOC will be the same in the end as in the beginning. All temporal losses will be "refilled" until the final time step.
- For the refill the flow defined in inputs will be used
- The self-discharging costs correspond with the costs related to the carrier at inputs (here: electricity).
- In the given example these costs are either a) variable-generation-costs + variable-storage-input-costs or b) variable-backup-costs + variable-storage-input-costs, depending on the origin of the electricity in the time step when the storage is charged.
- The given Excel-Sheet shows that all "storage_in" comes from "generation". That leads to costs-of-temporal-losses = variable-storage-input-costs, because generation-costs are zero. 
- In the Excel-Sheet the total system costs are calculated using a electriciy-loss-price of 6 while the actual electriciy-loss-price in oemof is zero, that of cause leads to different results.

In my opinion that is absolutly correct. If I balance my losses with electricity that is "for free" I should not expect costs. 
I don't see the need for pricing self-discharged electricity as this is already done via the `variable_costs` of the carrier. We should be aware that, in fact, this would lead to double pricing!

Hopefully, I could clearify what affects the discrepancy between the two total costs results.

**Comment by p-snft** [comment]
I do see a point in attributing a cost to losses. To be consistent, also the capacity at the first and last time step have to be taken into account. With `balanced=True`, these are of course equal.

Still, as @jakob-wo pointed out, the current way to model this is to add costs to consider the costs of the generation. Setting `variable_costs=-generation_costs` on the demand side and will yield the same result as you have when pricing discharge. Doing this, however, requires a balanced storage.

**Comment by baumconsult** [comment]
The key point might be formulated as follows: the costs must be attributed to the component which causes them. Otherwise the optimization will not find the optimum combination of components and the related dispatch.
Example: A storage leads to a somewhat higher energy need due to the losses occuring in the storage when operated. If the costs of these losses are not attributed to the storage, but to a variable source, the optimization process will come out with a storage whose size is too big.

**Comment by baumconsult** [comment]
To get some further insight, I have written three versions of the storage check programm where I now include investment cost also for the backup generation and I remove curtailment costs. The storyline is:
A fixed generation g is to match a fixed demand d by a combination of curtailment, backup generation or storage. Backup generation and storage both have investment costs and their capacities are determined by the optimisation.
storage_check_3a: variable costs of 1 per unit of backup generation are added to the investment costs.
storage_check_4a: variable costs of 1 per unit of energy lost in the storage is added to the investment costs.
storage_check_5a: variable costs of 1 per unit of backup generation and of energy lost in the storage is added to the investment costs.
All 3 cases deliver a different combination of optimum backup generation and storage. In the two cases where variable costs are attributed to the energy lost in the storage there is a difference between the expressively calculated total costs (same values obtained in oemof and in excel) and the objective value. In both cases the difference equals exactly the costs of energy lost by self-discharge in the storage.

**Question: Which of the 3 codes provides the real optimum solution?**

[storage_check_3a.txt](https://github.com/oemof/oemof/files/3306806/storage_check_3a.txt)
[storage_check_4a.txt](https://github.com/oemof/oemof/files/3306807/storage_check_4a.txt)
[storage_check_5a.txt](https://github.com/oemof/oemof/files/3306808/storage_check_5a.txt)
[storage_analysis_3a.xlsx](https://github.com/oemof/oemof/files/3306809/storage_analysis_3a.xlsx)
[storage_analysis_4a.xlsx](https://github.com/oemof/oemof/files/3306811/storage_analysis_4a.xlsx)
[storage_analysis_5a.xlsx](https://github.com/oemof/oemof/files/3306812/storage_analysis_5a.xlsx)

**Comment by baumconsult** [comment]
> 
> 
> I do not understand the general idea off adding loss costs. Electricity losses produce costs because one has to produce more energy to satisfy the demand but I would assign extra costs for the losses.

This might be true in case of a simple dispatch problem. If components are sized in the optimisation, my comment applies that costs must be attributed to the component which causes them.

**Comment by baumconsult** [comment]
> 
> 
> I do see a point in attributing a cost to losses. To be consistent, also the capacity at the first and last time step have to be taken into account. With `balanced=True`, these are of course equal.
> 
> Still, as @jakob-wo pointed out, the current way to model this is to add costs to consider the costs of the generation. Setting `variable_costs=-generation_costs` on the demand side and will yield the same result as you have when pricing discharge. Doing this, however, requires a balanced storage.

In addition to my last comment, I do not see where to include `variable_costs=-generation_costs` in my code. Also, as far as I know, oemof cannot handle negative numbers.

**Comment by baumconsult** [comment]
Reading the replies here, I wonder if you keep in mind that costs of electricity are always a product of specific costs per unit electricity (e.g. measured in €/kWh) and a certain amount of electricity (e.g. measured in kWh/timestep). In case of costs of lost energy, the latter must be the electricity which is lost when being moved, not the electricity which is moved. That is for instance the electricity charged to a storage * charging efficiency, not the electricity charged to the storage. See my code for the variable costs definition for the storage.

**Comment by p-snft** [comment]
> In addition to my last comment, I do not see where to include `variable_costs=-generation_costs` in my code. Also, as far as I know, oemof cannot handle negative numbers.

It can. The flows are defined to be positive, to have a clear direction, bit you can have negative costs (= revenue). See [storage_check_6.txt](https://github.com/oemof/oemof/files/3309738/storage_check_6.txt). This version also has total_costs equal to the optimal objective value (both 81.71).

Note that a bigger storage creates more losses and thus will impact the investment decision. Though, it is not straightforward to attribute the additional costs to the losses for more complex networks: They just emerge as the difference between in- and outflow. So, as I said, I do see a point in attributing a cost to losses.

**Comment by FranziPl** [comment]
> The key point might be formulated as follows: the costs must be attributed to the component which causes them. Otherwise the optimization will not find the optimum combination of components and the related dispatch.
As I understand it, it is important to attribute the costs to the correct component, but not to a specific loss. So input-, output- and temporal losses could be considered together. Wouldn't it be a possibilty to "pay" for the storage input with variable_costs = cl and "get paid" for the output with variable_costs = -cl. The costs for the losses are the difference between the costs of the two flows. In this way, objective value and total costs are the same.

**Comment by p-snft** [comment]
The code I posted was inconsistent, as it weights some losses twice. I have not seen the existing costs in the storage flows. Anyway, what @FranziPl suggested is better than what I did. Let me summarize:

```python
storage = energysystem.add(solph.components.GenericStorage(
            label='storage',
            inputs={bel: solph.Flow(variable_costs=cl*(2-icf))},
            outputs={bel: solph.Flow(variable_costs=cl*((1-ocf)/ocf)-1)},
            ...
```

As inflows-outflows-losses=0, this is equivalent to adding a price to the losses.

**Comment by baumconsult** [comment]
Thanks a lot. I think @FranziPl got it, but I would also have a closer look on @p-snft's comments. I will come back here when I have checked everything thoroughly.

**Comment by baumconsult** [comment]
Well .... checking things opened more questions that it answered ....

@p-snft : I could not confirm your findings. The objective value is different from the calculated total costs. See storage_check_6a.txt which shows a significant difference for a different parameter setting. Also by checking the formulae, I could not confirm that one to them reflects correctly the cost of losses in the storage.

@FranziPl : See storage_check_7.txt. Your approach certainly reproduces correctly the cost of losses in the storage. The objective value is equal to the calculated total costs.

-> Good messange: There is no need to modify the GenericStorage-Module, because it is possible to access the total losses and their costs in the way shown by @FranziPl.

But ... I belief also that @uvchik is right with his comment that losses in a storage appear somewhere else as costs of additional generation. Looking at storage_check_7.txt, I think:

- The objective value and the calculated total cost are NOT the real additional cost of the flexibility provided by the combination of curtailment, backup generation and storage. The real cost of that flexibility are the sum of the storage investment costs and the backup generation costs - here @uvchik is right.
- As mentionned earlier, the costs of the losses in the storage must be attributed to the storage. Otherwise, oemof does not find the optimum combination of components.
- We must somehow distinguish the objective value and the real costs of flexibility ....

So far for tonight ...

[storage_check_6a.txt](https://github.com/oemof/oemof/files/3315467/storage_check_6a.txt)
[storage_check_7.txt](https://github.com/oemof/oemof/files/3315468/storage_check_7.txt)

**Comment by p-snft** [comment]
I really like the enthusiasm you put into testing and I have to apologise for bringing confusion with my code above. The point I wanted to make is that negative costs are possible — but I did not check the other costs you applied. So, the script is inconsistent.

> As mentionned earlier, the costs of the losses in the storage must be attributed to the storage. Otherwise, oemof does not find the optimum combination of components.

Here, I have to disagree: When higher losses cause more cost, the size of the storage will be limited no matter where these costs appear. (Typically it will be in the generation.) As oemof finds the global optimum, it does not matter where they are attributed. However, it simplifies analysing the results when the costs are visibly connected to the component. (Actually, variable_costs are part of the flows, not of the components or busses they connect.)

**Comment by baumconsult** [comment]
By the way: Your comment that costs can be negative is extremely helpful!

**Comment by uvchik** [comment]
>    As mentionned earlier, the costs of the losses in the storage must be attributed to the storage. Otherwise, oemof does not find the optimum combination of components

Oemof uses the system costs to find the optimal combination. In an investment model the costs to build up a storage must be lower than the overall gain for the system with that storage. If the losses of a storage are very high and a lot of extra energy is needed to compensate the losses the storage might not be build and an alternative component might be preferred.

**Comment by baumconsult** [comment]
I think this issues is solved. I stick to uvchik with the solution to consider costs related to storage losses not in the storage component but in those components where the costs finally appear. In the example which I looked into in more detail, this is a backup generation component. I documented the check here: https://github.com/baumconsult/goflex
Thanks a lot for all your feedback. The other comments not considered now will be very helpful for further investigations.
